src/combinatorics/hales_jewett.lean

```diff
@@ -0,0 +1,313 @@
+Hales-Jewett theorem states that for any fixed finite types `α` and `κ`, there exists a (potentially
+huge) finite type `ι` such that whenever `ι → α` is `κ`-colored (i.e. for any coloring
+`C : (ι → α) → κ`), there exists a monochromatic line. We prove the Hales-Jewett theorem using
+the idea of *color focusing* and a *product argument*. See the proof of
+`combinatorics.line.exists_mono_in_high_dimension` for details.
+
+The version of Van der Waerden's theorem in this file states that whenever a commutative monoid `M`
+is finitely colored and `S` is a finite subset, there exists a monochromatic homothetic copy of `S`.
+This follows from the Hales-Jewett theorem by considering the map `(ι → S) → M` sending `v`
+to `∑ i : ι, v i`, which sends a combinatorial line to a homothetic copy of `S`.
+
+## Main results
+
+- `combinatorics.line.exists_mono_in_high_dimension`: the Hales-Jewett theorem.
+- `combinatorics.exists_mono_homothetic_copy`: a generalization of Van der Waerden's theorem.
+
+## Implementation details
+
+For convenience, we work directly with finite types instead of natural numbers. That is, we write
+`α, ι, κ` for (finite) types where one might traditionally use natural numbers `n, H, c`. This
+allows us to work directly with `α`, `option α`, `(ι → α) → κ`, and `ι ⊕ ι'` instead of `fin n`,
+`fin (n+1)`, `fin (c^(n^H))`, and `fin (H + H')`.
+
+## Todo
+
+- Prove a finitary version of Van der Waerden's theorem (either by compactness or by modifying the
+current proof).
+
+- Make the Hales-Jewett theorem universe-polymorphic, and maybe return `n : ℕ` instead of
+`ι : Type`?
+
+- One could reformulate the proof of Hales-Jewett to give explicit upper bounds on the number of
+coordinates needed.
+
+## Tags
+
+combinatorial line, Ramsey theory, arithmetic progession
+
+### References
+
+* https://en.wikipedia.org/wiki/Hales%E2%80%93Jewett_theorem
+
+-/
+
+open_locale classical
+open_locale big_operators
+
+namespace combinatorics
+
+/-- An induction principle for finite types. It effectively says that any finite type is equivalent
+to either `empty`, `unit`, or `option α` where `α` is nonempty. -/
+lemma fintype_induction {P : Type → Prop} (h_congr : ∀ {α β}, α ≃ β → P α → P β)
+  (h_empty : P empty) (h_unit : P unit)
+  (h_option : ∀ {α} [fintype α] [nonempty α], P α → P (option α))
+  (α) [fintype α] : P α :=
+begin
+  suffices : ∀ ι : nat, P (fin ι),
+  { refine h_congr (fintype.equiv_fin α).symm (this _), },
+  rintro (_ | ι),
+  { apply h_congr fin_zero_equiv.symm h_empty },
+  induction ι with ι ih,
+  { apply h_congr fin_one_equiv.symm h_unit },
+  { apply h_congr (fin_succ_equiv _).symm (h_option ih), }
+end
+
+/-- The type of combinatorial lines. A line `l : line α ι` in the hypercube `ι → α` defines a
+function `α → (ι → α)` to the hypercube, such that for each coordinate `i : ι`, the function
+`l x i` is either identically `x` or constant. We require lines to be nontrivial, in the sense
```

Comment:
I think "is `id`" or "is identically `id`" will be clearer than "is identically `x`". Also, if I understood correctly, `l x i` is **not** a function. I don't exactly know how to rewrite it, but it's currently very unclear.

Reply:
Sorry, by `l x i` I meant the function `λ x, l x i`. Is the new version clearer?